docs/data-pipeline.md

```diff
@@ -0,0 +1,160 @@
+
+### Static Data Slicing
+
+Static analysis of OmnICU code can determine which keys are required for a particular build.  This information can be used to automatically generate a minimal JSON file, for example.
+
+The examples `CURR_SYM_V1` and `CURR_LOCAL_SYM_V1` illustrate an important aspect of data slicing.  If the OmnICU code only ever formats the locale's preferred currency symbol, it can use `CURR_LOCAL_SYM_V1`, and the static data slicer can include only that small piece of data.  However, if OmnICU is compiled to accept any arbitrary currency code, then it should use `CURR_SYM_V1`, and the static data slicer knows that it must include symbols for all currencies.
+
+*Note: For the currency example specifically, this may also involve making OmnICU configurable for whether or not it accepts the `-u-cu-` Unicode extension in the locale string, since supporting that extension may require loading full currency data.*
+
+### Mappings
+
+The data on disk may not reflect exactly the types required for a particular key.  For example, imagine we have the following data file:
+
+```json
+{
+    "gregorian_months": [
+        "January", "February", "March", "April", "May", "June",
+        "July", "August", "September", "October", "November", "December"
+    ]
+}
+```
+
+The mapping is responsible for translating that schema into the hunk type required for a particular key.  For example, the pseudocode of a mapping might be:
+
+```
+switch (key) {
+    case DATE_SYM_JAN_V1:
+        return json["gregorian_months"][0];
+    case DATE_SYM_FEB_V1:
+        return json["gregorian_months"][1];
+    case DATE_SYM_MONTHS_V1:
+        return json["gregorian_months"];
+}
+```
+
+### Supported Keys
+
+A data provider need not implement a mapping for all keys.  For example, if OmnICU migrates from `NUM_SYM_GROUP_V1` to `NUM_SYM_GROUP_V2`, a data provider might choose to support both keys for a few releases, and at some point drop support for the older key.  This allows the same data provider to be used across multiple OmnICU versions.
+
+### Schema Version
+
+A data file (such as JSON) should represent data conforming to a certain schema, and the data provider's mapping should map from a specific schema version to a specific set of supported keys.  The data file can be rebuilt with different data versions so long as it conforms to the same schema version.
+
+The expectation is that an offline tool transforms CLDR data into a specific data schema version and writes out a data file that the data provider uses at runtime.  It is also possible that a data provider could read from CLDR's raw XML or JSON files directly, although this might not be as efficient at runtime.
+
+### Types
+
+The set of supported types is limited so that OmnICU implementations can be written to support them in a consistent way.  The supported types are:
+
+- i8
```

Comment:
Should we describe the types in language-independent ways?  Ex: signed 8-bit int (i8), signed 32-bit int (i32).  It wouldn't preclude you from using the Rust type signature as shorthand in your communications.

Reply:
I didn't intend for `i8` and `i32` to imply being rust-specific; I borrowed those names because they were more concise than writing out "8-bit signed integer" and "32-bit signed integer".